sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Filter.java

```diff
@@ -0,0 +1,186 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.schemas.transforms;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.Row;
+
+/**
+ * A {@link PTransform} for filtering a collection of schema types.
+ *
+ * <p>Separate Predicates can be registered for different schema fields, and the result is allowed
+ * to pass if all predicates return true. The output type is the same as the input type.
+ *
+ * <p>For example, consider the following schema type:
+ *
+ * <pre>{@code
+ * public class Location {
+ *   public double latitude;
+ *   public double longitude;
+ * }
+ * }</pre>
+ *
+ * In order to examine only locations in south Manhattan, you would write:
+ *
+ * <pre>{@code
+ * PCollection<Location> locations = readLocations();
+ * locations.apply(Filter
+ *    .whereFieldName("latitude", lat -> lat < 40.720 && lat > 40.699)
+ *    .whereFieldName("longitude", long -> long < -73.969 && long > -74.747));
+ * }</pre>
+ *
+ * Predicates that require examining multiple fields at once are also supported. For example,
+ * consider the following class representing a user account:
+ *
+ * <pre>{@code
+ * class UserAccount {
+ *   public double spendOnBooks;
+ *   public double spendOnMovies;
+ *       ...
+ * }
+ * }</pre>
+ *
+ * Say you want to examine only users whos total spend is above $100. You could write:
+ *
+ * <pre>{@code
+ * PCollection<UserAccount> users = readUsers();
+ * users.apply(Filter
+ *    .whereFieldNames(Lists.newArrayList("spendOnBooks", "spendOnMovies"),
+ *        row -> return row.getDouble("spendOnBooks") + row.getDouble("spendOnMovies") > 100.00));
+ * }</pre>
+ */
+@Experimental(Kind.SCHEMAS)
+public class Filter {
+  public static <T> Inner<T> create() {
+    return new Inner<T>();
+  }
+
+  /** Implementation of the filter. * */
+  public static class Inner<T> extends PTransform<PCollection<T>, PCollection<T>> {
+    Map<String, SerializableFunction<?, Boolean>> fieldNameFilters = Maps.newHashMap();
```

Comment:
These should probably all be `private final`.

Reply:
done